diff --git a/src/core/display.c b/src/core/display.c
index 45f8080bd3..6fe51fea9f 100644
--- a/src/core/display.c
+++ b/src/core/display.c
@@ -3556,7 +3556,7 @@ focus_on_pointer_rest_callback (gpointer data)
  * mode to check whether the pointer has stopped moving after a
  * crossing event.
  */
-#define FOCUS_TIMEOUT_DELAY 25
+#define FOCUS_TIMEOUT_DELAY 0
 
 static void
 queue_pointer_rest_callback (MetaDisplay *display,
