diff --git a/src/compositor/meta-window-drag.c b/src/compositor/meta-window-drag.c
index 00b0ce2797..b9a83d493d 100644
--- a/src/compositor/meta-window-drag.c
+++ b/src/compositor/meta-window-drag.c
@@ -2027,7 +2027,7 @@ meta_window_drag_begin (MetaWindowDrag      *window_drag,
   meta_topic (META_DEBUG_FOCUS,
               "Focusing %s because we're grabbing all its keys",
               window->desc);
-  meta_window_focus (window, timestamp);
+  meta_window_focus (window, timestamp, FALSE);
 
   stage = CLUTTER_STAGE (meta_backend_get_stage (backend));
 
diff --git a/src/core/display.c b/src/core/display.c
index 45f8080bd3..c6b8d50ead 100644
--- a/src/core/display.c
+++ b/src/core/display.c
@@ -1150,7 +1150,7 @@ on_is_grabbed_changed (ClutterStage *stage,
   if (!priv->focus_on_grab_dismissed.window)
     return;
 
-  meta_window_focus (priv->focus_on_grab_dismissed.window, META_CURRENT_TIME);
+  meta_window_focus (priv->focus_on_grab_dismissed.window, META_CURRENT_TIME, TRUE);
 
   g_clear_signal_handler (&priv->focus_on_grab_dismissed.unmanaging_handler_id,
                           priv->focus_on_grab_dismissed.window);
@@ -3484,7 +3484,7 @@ focus_mouse_mode (MetaDisplay *display,
       meta_topic (META_DEBUG_FOCUS,
                   "Focusing %s at time %u.", window->desc, timestamp_ms);
 
-      meta_window_focus (window, timestamp_ms);
+      meta_window_focus (window, timestamp_ms, TRUE);
 
       if (meta_prefs_get_auto_raise ())
         meta_display_queue_autoraise_callback (display, window);
diff --git a/src/core/meta-workspace-manager.c b/src/core/meta-workspace-manager.c
index c314bd8119..c7bc4ed737 100644
--- a/src/core/meta-workspace-manager.c
+++ b/src/core/meta-workspace-manager.c
@@ -963,7 +963,7 @@ meta_workspace_manager_show_desktop (MetaWorkspaceManager *workspace_manager,
 
       if (w->type == META_WINDOW_DESKTOP)
         {
-          meta_window_focus (w, timestamp);
+          meta_window_focus (w, timestamp, TRUE);
           break;
         }
     }
diff --git a/src/core/window.c b/src/core/window.c
index 3cc26a2520..1dffdca7ab 100644
--- a/src/core/window.c
+++ b/src/core/window.c
@@ -2589,7 +2589,7 @@ meta_window_show (MetaWindow *window)
           timestamp = meta_display_get_current_time_roundtrip (window->display);
 
           if (meta_display_windows_are_interactable (window->display))
-            meta_window_focus (window, timestamp);
+            meta_window_focus (window, timestamp, TRUE);
           else
             meta_display_queue_focus (window->display, window);
         }
@@ -3916,7 +3916,7 @@ meta_window_activate_full (MetaWindow     *window,
               window->desc);
 
   if (meta_window_located_on_workspace (window, workspace))
-    meta_window_focus (window, timestamp);
+    meta_window_focus (window, timestamp, TRUE);
   else
     meta_workspace_activate_with_focus (window->workspace, window, timestamp);
 
@@ -4998,7 +4998,8 @@ meta_window_make_most_recent (MetaWindow    *window,
 /* XXX META_EFFECT_FOCUS */
 void
 meta_window_focus (MetaWindow  *window,
-                   guint32      timestamp)
+                   guint32      timestamp,
+                   gboolean     warp)
 {
   MetaWorkspaceManager *workspace_manager = window->display->workspace_manager;
   MetaWindow *modal_transient;
@@ -5091,6 +5092,8 @@ meta_window_focus (MetaWindow  *window,
     meta_window_unset_demands_attention (window);
 
 /*  meta_effect_run_focus(window, NULL, NULL); */
+  if (warp)
+    meta_wayland_pointer_warp_window (window);
 }
 
 /* Workspace management. Invariants:
@@ -7736,7 +7739,7 @@ meta_window_handle_ungrabbed_event (MetaWindow         *window,
       meta_topic (META_DEBUG_FOCUS,
                   "Focusing %s due to button %u press (display.c)",
                   window->desc, button);
-      meta_window_focus (window, time_ms);
+      meta_window_focus (window, time_ms, FALSE);
       meta_window_check_alive (window, time_ms);
     }
 
diff --git a/src/core/workspace.c b/src/core/workspace.c
index 5ba623e8bf..832c1a4cfc 100644
--- a/src/core/workspace.c
+++ b/src/core/workspace.c
@@ -1420,7 +1420,7 @@ meta_workspace_focus_default_window (MetaWorkspace *workspace,
     {
       meta_topic (META_DEBUG_FOCUS,
                   "Focusing mouse window %s", window->desc);
-      meta_window_focus (window, timestamp);
+      meta_window_focus (window, timestamp, TRUE);
     }
 
   if (workspace->display->autoraise_window != window &&
@@ -1625,7 +1625,7 @@ try_to_set_focus_and_check (MetaWindow *window,
                             MetaWindow *not_this_one,
                             uint32_t    timestamp)
 {
-  meta_window_focus (window, timestamp);
+  meta_window_focus (window, timestamp, TRUE);
 
   /* meta_focus_window() will not change focus for clients using the
    * "globally active input" model of input handling, hence defeating
diff --git a/src/meta/window.h b/src/meta/window.h
index 761896b33e..85794c571a 100644
--- a/src/meta/window.h
+++ b/src/meta/window.h
@@ -371,7 +371,8 @@ void        meta_window_kill               (MetaWindow  *window);
 
 META_EXPORT
 void        meta_window_focus              (MetaWindow  *window,
-                                            guint32      timestamp);
+                                            guint32      timestamp,
+                                            gboolean     warp);
 
 META_EXPORT
 void        meta_window_check_alive        (MetaWindow  *window,
diff --git a/src/wayland/meta-wayland-pointer-warp.c b/src/wayland/meta-wayland-pointer-warp.c
index 89d9d1e7c8..efe7fa40e3 100644
--- a/src/wayland/meta-wayland-pointer-warp.c
+++ b/src/wayland/meta-wayland-pointer-warp.c
@@ -133,3 +133,17 @@ meta_wayland_pointer_warp_destroy (MetaWaylandPointerWarp *pointer_warp)
   wl_list_remove (&pointer_warp->resource_list);
   g_free (pointer_warp);
 }
+
+void
+meta_wayland_pointer_warp_window (MetaWindow *window) {
+  MetaDisplay *display = window->display;
+  MetaCompositor *compositor = meta_display_get_compositor (display);
+  MetaBackend *backend = meta_compositor_get_backend (compositor);
+  ClutterBackend *clutter_backend = meta_backend_get_clutter_backend (backend);
+  ClutterSeat *seat = clutter_backend_get_default_seat (clutter_backend);
+  MtkRectangle rect = window->buffer_rect;
+  int x = rect.x + rect.width / 2;
+  int y = rect.y + rect.height / 2;
+
+  clutter_seat_warp_pointer (seat, x, y);
+}
diff --git a/src/wayland/meta-wayland-pointer-warp.h b/src/wayland/meta-wayland-pointer-warp.h
index 658638b342..2948f7b836 100644
--- a/src/wayland/meta-wayland-pointer-warp.h
+++ b/src/wayland/meta-wayland-pointer-warp.h
@@ -24,6 +24,7 @@
 #include <glib.h>
 #include <wayland-server.h>
 
+#include "meta/types.h"
 #include "wayland/meta-wayland-types.h"
 
 typedef struct _MetaWaylandPointerWarp MetaWaylandPointerWarp;
@@ -31,3 +32,5 @@ typedef struct _MetaWaylandPointerWarp MetaWaylandPointerWarp;
 MetaWaylandPointerWarp * meta_wayland_pointer_warp_new (MetaWaylandSeat *seat);
 
 void meta_wayland_pointer_warp_destroy (MetaWaylandPointerWarp *pointer_warp);
+
+void meta_wayland_pointer_warp_window (MetaWindow *window);
diff --git a/src/wayland/meta-wayland-xdg-shell.c b/src/wayland/meta-wayland-xdg-shell.c
index d91b08a1af..5cb4f1ccec 100644
--- a/src/wayland/meta-wayland-xdg-shell.c
+++ b/src/wayland/meta-wayland-xdg-shell.c
@@ -1371,7 +1371,7 @@ finish_popup_setup (MetaWaylandXdgPopup *xdg_popup)
       MetaWaylandPopupSurface *popup_surface;
       MetaWaylandPopup *popup;
 
-      meta_window_focus (window, meta_display_get_current_time (display));
+      meta_window_focus (window, meta_display_get_current_time (display), TRUE);
       popup_surface = META_WAYLAND_POPUP_SURFACE (surface->role);
 
       if (!xdg_shell_client->popup_grab)
diff --git a/src/x11/events.c b/src/x11/events.c
index 96978dc82b..1689ae3fb8 100644
--- a/src/x11/events.c
+++ b/src/x11/events.c
@@ -1421,7 +1421,7 @@ handle_other_xevent (MetaX11Display *x11_display,
                window->reparents_pending == 0)
         {
           meta_window_focus (window,
-                             meta_display_get_current_time_roundtrip (display));
+                             meta_display_get_current_time_roundtrip (display), TRUE);
         }
 
       break;
diff --git a/src/x11/window-x11.c b/src/x11/window-x11.c
index d7dda58b7e..0893854b00 100644
--- a/src/x11/window-x11.c
+++ b/src/x11/window-x11.c
@@ -866,7 +866,7 @@ focus_window_delayed_timeout (gpointer user_data)
   data->timeout_id = 0;
   meta_window_x11_delayed_focus_data_free (data);
 
-  meta_window_focus (window, timestamp);
+  meta_window_focus (window, timestamp, TRUE);
 }
 
 static void
