diff --git a/src/core/window.c b/src/core/window.c
index 3cc26a2520..684bf5d67c 100644
--- a/src/core/window.c
+++ b/src/core/window.c
@@ -5578,8 +5578,7 @@ meta_window_update_appears_focused (MetaWindow *window)
   if (workspace && workspace != workspace_manager->active_workspace)
     {
       appears_focused =
-        window == meta_workspace_get_default_focus_window (workspace, NULL) &&
-        meta_prefs_get_focus_mode () == G_DESKTOP_FOCUS_MODE_CLICK;
+        window == meta_workspace_get_default_focus_window (workspace, NULL);
     }
   else
     {
