diff --git a/src/core/window.c b/src/core/window.c
index 3cc26a2520..30d657a01d 100644
--- a/src/core/window.c
+++ b/src/core/window.c
@@ -2588,10 +2588,13 @@ meta_window_show (MetaWindow *window)
 
           timestamp = meta_display_get_current_time_roundtrip (window->display);
 
-          if (meta_display_windows_are_interactable (window->display))
+          if (meta_display_windows_are_interactable (window->display)) {
+            // Open window
             meta_window_focus (window, timestamp);
-          else
+            pointer_warp_window (window);
+          } else {
             meta_display_queue_focus (window->display, window);
+          }
         }
     }
 
@@ -3915,10 +3918,13 @@ meta_window_activate_full (MetaWindow     *window,
               "Focusing window %s due to activation",
               window->desc);
 
-  if (meta_window_located_on_workspace (window, workspace))
+  if (meta_window_located_on_workspace (window, workspace)) {
+    // Focus another window
     meta_window_focus (window, timestamp);
-  else
+    pointer_warp_window (window);
+  } else {
     meta_workspace_activate_with_focus (window->workspace, window, timestamp);
+  }
 
   meta_display_maybe_update_sticky_mouse_focus (window->display);
 
diff --git a/src/core/workspace.c b/src/core/workspace.c
index 5ba623e8bf..d8aabd45bb 100644
--- a/src/core/workspace.c
+++ b/src/core/workspace.c
@@ -46,6 +46,7 @@
 #include "core/meta-workspace-manager-private.h"
 #include "core/workspace-private.h"
 #include "meta/prefs.h"
+#include "wayland/meta-wayland-pointer-warp.h"
 
 void meta_workspace_queue_calc_showing   (MetaWorkspace *workspace);
 static void focus_ancestor_or_mru_window (MetaWorkspace *workspace,
@@ -1625,7 +1626,9 @@ try_to_set_focus_and_check (MetaWindow *window,
                             MetaWindow *not_this_one,
                             uint32_t    timestamp)
 {
+  // Close window
   meta_window_focus (window, timestamp);
+  pointer_warp_window (window);
 
   /* meta_focus_window() will not change focus for clients using the
    * "globally active input" model of input handling, hence defeating
diff --git a/src/wayland/meta-wayland-pointer-warp.c b/src/wayland/meta-wayland-pointer-warp.c
index 89d9d1e7c8..df725a9c0a 100644
--- a/src/wayland/meta-wayland-pointer-warp.c
+++ b/src/wayland/meta-wayland-pointer-warp.c
@@ -35,6 +35,20 @@ struct _MetaWaylandPointerWarp
   struct wl_list resource_list;
 };
 
+void
+pointer_warp_window (MetaWindow *window) {
+  MetaDisplay *display = window->display;
+  MetaCompositor *compositor = meta_display_get_compositor (display);
+  MetaBackend *backend = meta_compositor_get_backend (compositor);
+  ClutterBackend *clutter_backend = meta_backend_get_clutter_backend (backend);
+  ClutterSeat *seat = clutter_backend_get_default_seat (clutter_backend);
+  MtkRectangle rect = window->buffer_rect;
+  int x = rect.x + rect.width / 2;
+  int y = rect.y + rect.height / 2;
+
+  clutter_seat_warp_pointer (seat, x, y);
+}
+
 static void
 pointer_warp_destroy (struct wl_client   *client,
                       struct wl_resource *resource)
diff --git a/src/wayland/meta-wayland-pointer-warp.h b/src/wayland/meta-wayland-pointer-warp.h
index 658638b342..aa8266f066 100644
--- a/src/wayland/meta-wayland-pointer-warp.h
+++ b/src/wayland/meta-wayland-pointer-warp.h
@@ -24,10 +24,13 @@
 #include <glib.h>
 #include <wayland-server.h>
 
+#include "meta/types.h"
 #include "wayland/meta-wayland-types.h"
 
 typedef struct _MetaWaylandPointerWarp MetaWaylandPointerWarp;
 
+void pointer_warp_window (MetaWindow *window);
+
 MetaWaylandPointerWarp * meta_wayland_pointer_warp_new (MetaWaylandSeat *seat);
 
 void meta_wayland_pointer_warp_destroy (MetaWaylandPointerWarp *pointer_warp);
